src/Servers/HttpSys/test/FunctionalTests/RequestHeaderTests.cs

```diff
@@ -174,6 +182,407 @@ public async Task RequestHeaders_ClientSendTransferEncodingAndContentLength_Cont
         }
     }
 
+    [ConditionalFact]
+    public async Task RequestHeaders_AllKnownHeadersKeys_Received()
+    {
+        string customHeader = "X-KnownHeader";
+        using var server = Utilities.CreateHttpServer(out var address, httpContext =>
+        {
+            var requestHeaders = httpContext.Request.Headers;
+            Assert.Equal(3, requestHeaders.Count);
+            Assert.Equal(requestHeaders.Keys.Count, requestHeaders.Count);
+            Assert.Equal(requestHeaders.Count, requestHeaders.Values.Count);
+            Assert.Contains(customHeader, requestHeaders.Keys);
+            Assert.Contains(requestHeaders[customHeader].First(), requestHeaders.Keys);
+            Assert.Contains(HeaderNames.Host, requestHeaders.Keys);
+            return Task.FromResult(0);
+        });
+
+        foreach ((HttpSysRequestHeader Key, string Value) testRow in HeaderTestData())
+        {
+            var key = testRow.Key.ToString();
+            var headerDictionary = new Dictionary<string, string>
+            {
+                { key, testRow.Value },
+                { customHeader, key }
+            };
+            await SendRequestAsync(address, headerDictionary);
+        }
+    }
+
+    [ConditionalFact]
+    public async Task RequestHeaders_AllUnknownHeadersKeys_Received()
+    {
+        using var server = Utilities.CreateHttpServer(out var address, httpContext =>
+        {
+            var requestHeaders = httpContext.Request.Headers;
+            Assert.Equal(4, requestHeaders.Count);
+            Assert.Equal(requestHeaders.Keys.Count, requestHeaders.Count);
+            Assert.Equal(requestHeaders.Count, requestHeaders.Values.Count);
+            Assert.Contains("X-UnknownHeader-0", requestHeaders.Keys);
+            Assert.Contains("My-UnknownHeader-1", requestHeaders.Keys);
+            Assert.Contains("X-UnknownHeader-2", requestHeaders.Keys);
+            Assert.Contains(HeaderNames.Host, requestHeaders.Keys);
+            return Task.FromResult(0);
+        });
+
+        var headerDictionary = new Dictionary<string, string>
+        {
+            { "X-UnknownHeader-0", "0" },
+            { "My-UnknownHeader-1", "1" },
+            { "X-UnknownHeader-2", "2" }
+        };
+        await SendRequestAsync(address, headerDictionary);
+    }
+
+    [ConditionalFact]
+    public unsafe void RequestHeaders_PooledMemory_ReturnsKnownHeadersKeys()
```

Comment:
Nice tests. Can you move any that don't touch the network over to 
https://github.com/dotnet/aspnetcore/tree/main/src/Servers/HttpSys/test/Tests

They might not need to be conditional either.